server/service/async/async.go

```diff
@@ -0,0 +1,281 @@
+
+	// log stats at regular intervals
+	// TODO: TBD if it stays that way for prod, but will be useful for load testing
+	go func() {
+		tick := time.Tick(time.Minute)
+		for {
+			select {
+			case <-tick:
+				stats := labelColl.ReadStats()
+				level.Debug(logger).Log("stats", fmt.Sprintf("%#v", stats), "name", labelColl.name)
+			case <-ctx.Done():
+				return
+			}
+		}
+	}()
+}
+
+func (t *Task) RecordLabelQueryExecutions(ctx context.Context, host *fleet.Host, results map[uint]*bool, ts time.Time) error {
+	if !t.AsyncEnabled {
+		host.Modified = true
+		host.LabelUpdatedAt = ts
+		return t.Datastore.RecordLabelQueryExecutions(ctx, host, results, ts)
+	}
+
+	keySet := fmt.Sprintf(labelMembershipHostKey, host.ID)
+	keyTs := fmt.Sprintf(labelMembershipReportedKey, host.ID)
+
+	script := redigo.NewScript(2, `
+    redis.call('ZADD', KEYS[1], unpack(ARGV, 2))
+    return redis.call('SET', KEYS[2], ARGV[1])
+  `)
+
+	// convert results to ZADD arguments, store as -1 for delete, +1 for insert
+	args := make(redigo.Args, 0, 3+(len(results)*2))
+	args = args.Add(keySet, keyTs, ts.Unix())
+	for k, v := range results {
+		score := -1
+		if v != nil && *v {
+			score = 1
+		}
+		args = args.Add(score, k)
+	}
+
+	conn := t.Pool.Get()
+	defer conn.Close()
+	if err := redis.BindConn(t.Pool, conn, keySet, keyTs); err != nil {
+		return errors.Wrap(err, "bind redis connection")
+	}
+
+	if _, err := script.Do(conn, args...); err != nil {
+		return err
+	}
+	return nil
+}
+
+var (
+	reHostFromKey = regexp.MustCompile(`\{(\d+)\}$`)
+
+	// those are variables so they can be set differently for tests
+	insertBatchSize = 2000
+	deleteBatchSize = 2000
+	redisPopCount   = 1000
+)
+
+func collectLabelQueryExecutions(ctx context.Context, ds fleet.Datastore, pool fleet.RedisPool, stats *collectorExecStats) error {
+	keys, err := redis.ScanKeys(pool, labelMembershipHostKeyPattern, 1000)
+	if err != nil {
+		return err
+	}
+	stats.Keys = len(keys)
+
+	getKeyTuples := func(key string) (inserts, deletes [][2]uint, err error) {
+		var hostID uint
+		if matches := reHostFromKey.FindStringSubmatch(key); matches != nil {
+			id, err := strconv.ParseInt(matches[1], 10, 64)
+			if err == nil && id > 0 && id <= math.MaxUint32 { // required for CodeQL vulnerability scanning in CI
+				hostID = uint(id)
+			}
+		}
+
+		// just ignore if there is no valid host id
+		if hostID == 0 {
+			return nil, nil, nil
+		}
+
+		conn := pool.ConfigureDoer(pool.Get())
+		defer conn.Close()
+
+		for {
+			vals, err := redigo.Ints(conn.Do("ZPOPMIN", key, redisPopCount))
+			if err != nil {
+				return nil, nil, errors.Wrap(err, "redis ZPOPMIN")
+			}
+			items := len(vals) / 2 // each item has the label id and the score (-1=delete, +1=insert)
+			stats.Items += items
+
+			for i := 0; i < len(vals); i += 2 {
+				labelID := vals[i]
+
+				var score int
+				if i+1 < len(vals) { // just to be safe we received all pairs
+					score = vals[i+1]
+				}
+
+				switch score {
+				case 1:
```

Comment:
No, it will always be 1 (insert) or -1 (delete), which is why the 0 case (and any other score) is ignored as it indicates an invalid member. `ZADD`, despite its name I guess :), adds or updates the member and its score, it doesn't do arithmetics on the score it just sets it to what is passed, so if multiple label results are reported for the same host before the collector can run, it will simply update any existing one (with a score of 1 or -1), which is exactly what we want (we just skipped an intermediate database insert or delete and saved unnecessary work).